go/genkit/reflection.go

```diff
@@ -256,6 +255,85 @@ func wrapReflectionHandler(h func(w http.ResponseWriter, r *http.Request) error)
 	}
 }
 
+// resolveAction tries to resolve any type of action and the dependant actions of it
+func resolveAction(g *Genkit, key string, input json.RawMessage) (core.Action, error) {
+	var atype, provider, name string
+
+	trimmedKey := strings.TrimPrefix(key, "/")
+	found := strings.HasPrefix(trimmedKey, string(core.ActionTypeUtil))
+
+	// special case when the action "util/generate" gets called with a model that has not been defined
+	if found && strings.Contains(key, "generate") {
+		var inputMap map[string]any
+		err := json.Unmarshal(input, &inputMap)
+		if err != nil {
+			return nil, core.NewError(core.INTERNAL, err.Error())
+		}
+		modelName, ok := inputMap["model"].(string)
+		if !ok {
+			if k, ok := inputMap["key"].(string); ok {
+				modelName, found = strings.CutPrefix(k, "model/")
+				if !found {
+					return nil, core.NewError(core.INVALID_ARGUMENT, "unable to get model name for action %q", inputMap["key"])
+				}
+			}
+		}
+		provider, name, found = strings.Cut(modelName, "/")
+		if !found {
+			return nil, core.NewError(core.INVALID_ARGUMENT, "unable to get provider from %q", modelName)
+		}
+		plugin := g.reg.LookupPluginByName(provider)
+		if plugin == nil {
+			return nil, core.NewError(core.NOT_FOUND, "plugin for provider %q not found", provider)
+		}
+		dp, ok := plugin.(DynamicPlugin)
+		if ok {
+			if a := g.reg.LookupAction(fmt.Sprintf("/%s/%s/%s", core.ActionTypeModel, provider, name)); a != nil {
+				return g.reg.LookupAction(key).(core.Action), nil
+			}
+			err = dp.ResolveAction(g, core.ActionTypeModel, name)
+			if err != nil {
+				return nil, core.NewError(core.INTERNAL, err.Error())
+			}
+		}
+
+		action := g.reg.LookupAction(key).(core.Action)
+		return action, nil
+	}
+
+	parts := strings.Split(trimmedKey, "/")
+	if len(parts) == 3 {
+		atype = parts[0]
+		provider = parts[1]
+		name = parts[2]
+
+		for _, plugin := range g.reg.ListPlugins() {
+			dp, ok := plugin.(DynamicPlugin)
+			if !ok {
+				continue
+			}
+			if dp.Name() != provider {
+				continue
+			}
+			if a := g.reg.LookupAction(fmt.Sprintf("/%s/%s/%s", atype, provider, name)); a != nil {
+				return g.reg.LookupAction(key).(core.Action), nil
+			}
```

Comment:
```suggestion
			if a := g.reg.LookupAction(fmt.Sprintf("/%s/%s/%s", atype, provider, name)); a != nil {
				return a.(core.Action), nil
			}
```

Is this what you meant? Doesn't make sense as is.

Reply:
You are right.